Features/update conviva to v4.x

sinewave440hz opened the pull request:
The changes in summary for the users of this sdk are:
1. The ConvivaAnalytics class is now called ConvivaAnalyticsIntegration
2. The third constructor `ConvivaAnalytics(Player player,
                            String customerKey,
                            Context context,
                            ConvivaConfig config,
                            Client client)`
 is now removed because the Client class is deprecated
 
 Other than that, the interface remains the same for now.

dweinber commented:
Please make sure to add a CHANGLOG.md entry, and I'd assume we need to update the build.gradle as well?
https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/blob/2e8e5b39a5aab839fbca791addd80e0f0e51b30c/conviva/build.gradle#L40

sinewave440hz commented:
> Please make sure to add a CHANGLOG.md entry, and I'd assume we need to update the build.gradle as well?
> 
> https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/blob/2e8e5b39a5aab839fbca791addd80e0f0e51b30c/conviva/build.gradle#L40

Yes, that would be good. The legacy SDKs were removed in 4.0.18. I'll test out the latest - v4.0.20.

sinewave440hz commented:
> > Please make sure to add a CHANGLOG.md entry, and I'd assume we need to update the build.gradle as well?
> > https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/blob/2e8e5b39a5aab839fbca791addd80e0f0e51b30c/conviva/build.gradle#L40
> 
> Yes, that would be good. The legacy SDKs were removed in 4.0.18. I'll test out the latest - v4.0.20.

This does, however, present us with a problem. The bitmovin integration's interface exposes objects from the legacy SDKs - the Client object. It's a simple enough change, but will change the interface for existing customers right away. This was always going to happen of course, but I rather expected there to be a transition phase. Should we dive in right away with that change?

sinewave440hz commented:
> > > Please make sure to add a CHANGLOG.md entry, and I'd assume we need to update the build.gradle as well?
> > > https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/blob/2e8e5b39a5aab839fbca791addd80e0f0e51b30c/conviva/build.gradle#L40
> > 
> > 
> > Yes, that would be good. The legacy SDKs were removed in 4.0.18. I'll test out the latest - v4.0.20.
> 
> This does, however, present us with a problem. The bitmovin integration's interface exposes objects from the legacy SDKs - the Client object. It's a simple enough change, but will change the interface for existing customers right away. This was always going to happen of course, but I rather expected there to be a transition phase. Should we dive in right away with that change?
> 
> (The alternative being to stay lower than 4.0.18 for that transition phase...)

A bit more about this - upgrading to 4.0.17.197 (immediately prior to the removal of the old SDK stuff) results in the Client class being marked as deprecated. So if there are any customers that are using the full interface including the Client parameter, perhaps that would be the best user experience for them - that their current interface works but is marked deprecated, and the fix is simply to remove the Client parameter.

dweinber commented:
Let's upgrade anyway. As you said, we'll need to do this at some point anyway. We'll manage Comms to clients about this change

sinewave440hz commented:
> Let's upgrade anyway. As you said, we'll need to do this at some point anyway. We'll manage Comms to clients about this change

Cool, done that. In that case, perhaps we should go ahead and make a change that I was also holding off on and that we discussed earlier - ie. changing the name of the Bitmovin ConvivaAnalytics class to ConvivaAnalyticsIntegration, so that we can use the unqualified name ConvivaAnalytics for Conviva without clashes? :)

dweinber commented:
Agreed, let's do that. Let's break multiple things at once instead of breaking one thing in every release 😉

dweinber commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -649,30 +516,15 @@ public void onEvent(PlayerEvent.TimeShifted timeShiftedEvent) {
-        } catch (ConvivaException e) {
-            Log.e(TAG, e.getLocalizedMessage());
-        }
+        // TODO - Verify...Conviva docs indicate SEEK_START rather than SEEK_STARTED...?
```
As far as I can see, `SEEK_START` doesn't exist in `ConvivaSdkConstants`, so should just be an error in the documentation, right?

dweinber commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -649,30 +516,15 @@ public void onEvent(PlayerEvent.TimeShifted timeShiftedEvent) {
-        } catch (ConvivaException e) {
-            Log.e(TAG, e.getLocalizedMessage());
-        }
+        // TODO - Verify...Conviva docs indicate SEEK_END rather than SEEK_ENDED...?
```
As far as I can see, `SEEK_END` doesn't exist in `ConvivaSdkConstants`, so should just be an error in the documentation, right?

sinewave440hz commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -12,88 +12,61 @@
-            SystemFactory androidSystemFactory = new SystemFactory(androidSystemInterface, systemSettings);
-            ClientSettings clientSettings = new ClientSettings(customerKey);
-
+        if(BuildConfig.DEBUG) {
```
Yes, see previous comment 👍

sinewave440hz commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -102,27 +75,17 @@ public void sendCustomApplicationEvent(String name) {
-        } catch (ConvivaException e) {
-            Log.e(TAG, e.getLocalizedMessage());
-        }
+        Log.d(TAG, "Will report app event: " + name + " " + attributes.toString());
```
It seems a little odd at first, but in fact session handling is now done behind the scenes upon calling the `init()`of the global `ConvivaAnalytics`class.

sinewave440hz commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -12,88 +12,61 @@
 
-    private void ensureConvivaSessionIsCreatedAndInitialized() {
-        if (!isSessionActive()) {
-            internalInitializeSession();
```
Hmm. It's a good point to consider. Although sessions are now handled internally in Conviva, I did see that having non-matching - that is, an unequal amount of - calls to `convivaVideoAnalytics.reportPlaybackRequested(contentMetadataBuilder.build());` and `convivaVideoAnalytics.reportPlaybackEnded();` results in Touchstone errors. This is why I reinstated a simpler form of session tracking, simply to track these methods and ensure they are "balanced". 

Having said that, you are quite right and I'm thinking that we should also reinstate that method so that we don't require the explicit `initializeSession()`. So I'll get on that!

dweinber commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -12,88 +12,61 @@
+            Map<String, Object> settings = new HashMap<String, Object>();
             if (config.getGatewayUrl() != null) {
-                clientSettings.gatewayUrl = config.getGatewayUrl();
+                settings.put(ConvivaSdkConstants.GATEWAY_URL, config.getGatewayUrl());
```
Ah okay, didn't expect that for the GATEWAY_URL. Fine for me then 👍

sinewave440hz commented:
All of the above discussions are now resolved so I'm requesting review on this once again :)

luckygoyal-bitmovin commented on README.md:
```diff
@@ -26,11 +26,11 @@ allprojects {
-  implementation 'com.conviva.sdk:conviva-core-sdk:2.145.1' // <-- conviva sdk
-  implementation 'com.bitmovin.analytics:conviva:1.1.1'
+  implementation 'com.conviva.sdk:conviva-core-sdk:4.0.20' // <-- conviva sdk
+  implementation 'com.bitmovin.analytics:conviva:2.0.2'
```
Update this file to replace `ConvivaAnalytics` with `ConvivaAnalyticsIntegration`

sinewave440hz replied:
This is done in [this commit](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/38/commits/80618197309f0e3441dd5254221fcfac33f25d84)

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ContentMetadataBuilder.java:
```diff
@@ -43,44 +44,53 @@ public void setPlaybackStarted(boolean value) {
                     metadataOverrides.getStreamType(),
                     metadata.getStreamType());
+            boolean isLive = streamType == ConvivaSdkConstants.StreamType.LIVE;
+            contentInfo.put(ConvivaSdkConstants.IS_LIVE, isLive);
```
`ConvivaSdkConstants.IS_LIVE` is not set in original code, any specific reason to add it now?

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -181,24 +135,23 @@ public void endSession() {
         this.updateSession();
     }
 
+    public void release() {
```
Please add a comment to explain usage of this public API

sinewave440hz replied:
This is done in [this commit](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/38/commits/80618197309f0e3441dd5254221fcfac33f25d84)

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -181,24 +135,23 @@ public void endSession() {
 
+    public void release() {
+        convivaVideoAnalytics.release();
+        ConvivaAnalytics.release();
```
Is there a need to release convivaAdAnalytics object as well?

sinewave440hz replied:
Ah yes, we're not actually using AdAnalytics at the moment so that should probably be removed for now.

sinewave440hz commented on CHANGELOG.md:
```diff
@@ -4,6 +4,17 @@ All notable changes to this project will be documented in this file.
+### [2.0.2]
+### Added
+
+- This version features an update to the Conviva SDK to 4.0.20. This removes the old V3 SDK, which means that any users that make use of the constructor including the Client parameter will now need to simply remove that Client parameter.
```
Ok, I'll update that 👍

sinewave440hz replied:
This is done in [this commit](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/pull/38/commits/80618197309f0e3441dd5254221fcfac33f25d84)

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -230,95 +176,102 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
-        } catch (ConvivaException e) {
-            Log.e(TAG, e.getLocalizedMessage());
+        if(!isSessionActive) {
+            convivaVideoAnalytics.reportPlaybackRequested();
```
Same as comment for pauseTracking(), does `reportPlaybackRequested` not start a new session instead of restarting monitoring of existing session?

sinewave440hz replied:
Well, we don't actually know how sessions are now handled, behind the scenes. The bitmovin term for session is no longer a match for the Conviva concept. I'm trying to get more info about this than the docs offer (from our Conviva contacts), but this what the discussions above in this PR were about. In this case, we are simply ensuring that the reportPlaybackRequested/reportPlaybackEnded calls are "balanced". Without this, touchstone reports errors basically.

luckygoyal-bitmovin replied:
thanks @sinewave440hz . I understand your point about sessions but my concern here is more about parity between stopping and resuming monitoring in a session(old code) VS terminating and starting a new session(new code). Is it possible to run a test using pauseTracking() and resumeTracking() APIs with old code and compare the touchstone session report with another session run using new code to check if the outcome looks similar?

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -230,95 +176,102 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
+     */
+    public void reportAppForegrounded() {
+        Log.d(TAG, "appForegrounded");
+        ConvivaAnalytics.reportAppForegrounded();
```
`isSessionActive` is set to `false` in reportAppBackgrounded but is not set to `true` again here. Is this expected? What about the use case when application is backgrounded thus marking session as inactive and then user foregrounds app resuming playback?

sinewave440hz replied:
I implemented it this way because it was what worked in touchstone, after some trial and error. But again, I have to research the app backgrounding issue.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -433,11 +365,8 @@ private void trackAdEnd() {
-            Log.e(TAG, e.getLocalizedMessage());
-        }
+        Log.d(TAG, "Report ad break ended");
+        convivaVideoAnalytics.reportAdBreakEnded();
```
Are there separate APIs for reporting AdBreakEnd and AdEnd? The trackAdEnd() method is being called when AdEnd event is received from Bitmovin player instead of when AdBreakEnd is received. If there are separate APIs, then we need to align the calls with respective Ad events from player.

sinewave440hz replied:
It's unfortunate that they are called different things, but they are the only calls available and actually lead to more detailed reporting than the previous version.

luckygoyal-bitmovin replied:
ok, thanks.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -681,14 +569,8 @@ public void setSeekEnd() {
-            } catch (ConvivaException e) {
-                Log.e(TAG, e.getLocalizedMessage());
-            }
+            convivaVideoAnalytics.reportAdBreakStarted(ConvivaSdkConstants.AdPlayer.CONTENT, ConvivaSdkConstants.AdType.SERVER_SIDE);
```
Same comment as AdTrackEnd.
Also, should this use  ConvivaSdkConstants.AdType.CLIENT_SIDE instead? This AdStarted event is fired only for client side Ads.

sinewave440hz replied:
Yes. Changed that.

luckygoyal-bitmovin replied:
thanks. Will resolve this when the related code change is committed.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -230,95 +176,102 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
-                playerStateManager.setBitrateKbps(bitrate);
-                playerStateManager.setVideoHeight(videoQuality.getHeight());
-                playerStateManager.setVideoWidth(videoQuality.getWidth());
-                clientMeasureInterface.setFrameRate(videoQuality.getFrameRate());
```
Since now ClientMeasureInterface is removed, how is frame rate information updated for the playback session?

sinewave440hz replied:
This info is actually still sent in `updateSession()`: 
```
        VideoQuality videoQuality = bitmovinPlayer.getPlaybackVideoData();
        if (videoQuality != null) {
            int bitrate = videoQuality.getBitrate() / 1000; // in kbps
                convivaVideoAnalytics.reportPlaybackMetric(ConvivaSdkConstants.PLAYBACK.RESOLUTION, videoQuality.getHeight(), videoQuality.getWidth());
                convivaVideoAnalytics.reportPlaybackMetric(ConvivaSdkConstants.PLAYBACK.BITRATE, bitrate);
                convivaVideoAnalytics.reportPlaybackMetric(ConvivaSdkConstants.PLAYBACK.RENDERED_FRAMERATE, Math.round(videoQuality.getFrameRate()));
        }
```

luckygoyal-bitmovin replied:
Ah..thanks. Sorry missed that earlier.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -12,88 +12,61 @@
-    private PlayerStateManager playerStateManager;
-    private ClientMeasureInterface clientMeasureInterface;
+    private ConvivaVideoAnalytics convivaVideoAnalytics;
+    private ConvivaAdAnalytics convivaAdAnalytics;
```
`ConvivaAdAnalytics` object is only created but is not used anywhere in the code? Should this be used for providing Ad related metadata?

luckygoyal-bitmovin replied:
ok, for my clarity, do you mean that Bitmovin team did not request for Ad reporting to be supported in this upgrade  and was supposed to be taken up as a follow-up change? Sorry for asking but I may be missing some information about the scope:)

sinewave440hz replied:
Nothing specific was discussed here, but the focus was of course on maintaining the same level of reporting as in the previous version of the SDK. The changes in the latest version handle some of the ad related stuff automatically, but I will check again that we are not missing anything that was available previously.

sinewave440hz replied:
We are not missing anything, and I suggest that we implement Ad Analytics at a later date.

luckygoyal-bitmovin replied:
ok, thanks. To be sure I understood correctly, the updated implementation is at par with existing implementation for Ad reporting, correct?

sinewave440hz replied:
Yes, that's correct. Tried and tested in touchstone too.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -681,14 +569,8 @@ public void setSeekEnd() {
         @Override
         public void onEvent(PlayerEvent.AdStarted adStartedEvent) {
             Log.d(TAG, "[Player Event] AdStarted");
-            Client.AdPosition adPosition = AdEventUtil.parseAdPosition(adStartedEvent, bitmovinPlayer.getDuration());
```
Is this Ad metadata like Ad position/duration etc. needed to passed into ConvivaAdAnalytics?

sinewave440hz replied:
This is not needed at present.

luckygoyal-bitmovin replied:
ok. thanks.

luckygoyal-bitmovin commented:
- Test cases(and associated app) need to be updated(Already discussed earlier in Slack)
- [ExampleApp](https://github.com/bitmovin/bitmovin-player-android-analytics-conviva/tree/main/ConvivaExampleApp) also needs to be updated to use `ConvivaAnalyticsIntegration` instead of `ConvivaAnalytics`.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ContentMetadataBuilder.java:
```diff
@@ -43,44 +44,53 @@ public void setPlaybackStarted(boolean value) {
+            }
 
-            contentMetadata.custom = getCustom();
+            contentInfo.putAll(getCustom());
```
Ah, thanks. Sorry, totally missed that but in hind sight it looks obvious:)

sinewave440hz commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -230,95 +176,102 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
+        Log.d(TAG, "Will report ad break started: " + ConvivaSdkConstants.AdPlayer.SEPARATE + " " + ConvivaSdkConstants.AdType.CLIENT_SIDE);
+        convivaVideoAnalytics.reportAdBreakStarted(ConvivaSdkConstants.AdPlayer.SEPARATE, ConvivaSdkConstants.AdType.CLIENT_SIDE);
+        if(isSessionActive) {
+            convivaVideoAnalytics.reportPlaybackEnded();
```
I'm not entirely sure tbh, I will investigate. In this case, I tried to find the equivalent behaviour according to the migration docs. Do you have an example of when/how pause/resumeTracking are used?

luckygoyal-bitmovin replied:
This implementation was based on a following recommentation by Conviva https://community.conviva.com/site/global/platforms/other_platforms/sdks/ios_sdk/tutorials/integratecustom/old-index.gsp#handling_ads

Unfortunately, now we do not have access to above documentation for iOS or Android but if possible you can look for this in Conviva's old Android SDK documentation.

sinewave440hz commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -230,95 +176,102 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
+     */
+    public void reportAppBackgrounded() {
+        Log.d(TAG, "appBackgrounded");
+        if(isSessionActive) {
```
Need to investigate this a bit more.

luckygoyal-bitmovin replied:
thanks

sinewave440hz replied:
This is no longer the case with the latest changes - backgrounding and foregrounding are now independent of session status, as it should be.

luckygoyal-bitmovin replied:
Thanks. Will resolve this when the related code change is committed.

sinewave440hz commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -230,95 +176,102 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
+        Log.d(TAG, "appBackgrounded");
+        if(isSessionActive) {
+            ConvivaAnalytics.reportAppBackgrounded();
+            isSessionActive = false;
```
I'll investigate the background playback case a little more to answer this...

luckygoyal-bitmovin replied:
thanks

luckygoyal-bitmovin replied:
Thanks for looking into this. In general applications can perform playback even when backgrounded so ideally all playback controls(pause, resume, seek and events etc.) should be allowed in background mode. With the current logic of appBackGrounded and appForeGrounded logic, the playback session status is tied to app background/foreground status. IMO these are independent. `isSessionActive` seems to be used here to check the last application state (backgrounded or foregrounded) instead of the session state. Should we use a different variable like `isBackgrounded` instead?

sinewave440hz replied:
Yes, that makes sense. I'll check that in with the next changes.

luckygoyal-bitmovin commented on conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalyticsIntegration.java:
```diff
@@ -210,16 +163,9 @@ public void reportPlaybackDeficiency(String message, Client.ErrorSeverity severi
-        }
+    public void reportPlaybackDeficiency(String message, ConvivaSdkConstants.ErrorSeverity severity, Boolean endSession) {
+        Log.d(TAG, "Will report playback deficiency: " + message + ",  " + severity.toString());
+        convivaVideoAnalytics.reportPlaybackError(message, severity);
```
thanks

sinewave440hz replied:
After testing this by hooking into the method in the example app, I realise that the best way to handle it is indeed to only send these if a session is active. The endSession part of the API can remain the same, I've tested that successfully.

luckygoyal-bitmovin commented:
@sinewave440hz , thanks for addressing review comments. The change look good to me now. IMO, the only pending work in updating the test cases to work again. Do you intend to work on test cases now or later?

sinewave440hz commented:
> @sinewave440hz , thanks for addressing review comments. The change look good to me now. IMO, the only pending work in updating the test cases to work again. Do you intend to work on test cases now or later?

I had started to look at these actually. However, I did notice that they don't seem to be passing now in master either. This may mean that it will take longer than my original half day estimate, I'm afraid, as I don't appear to have working code to refer to. Unless perhaps you can see what is wrong, or indeed confirm that you are seeing the same failures.
